models/demos/t3000/llama2_70b/tests/test_llama_generation.py

```diff
@@ -0,0 +1,177 @@
+# SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+
+# SPDX-License-Identifier: Apache-2.0
+import pytest
+from loguru import logger
+import torch
+from torch import nn
+import tt_lib
+import ttnn
+from ttnn import ShardTensorToMesh, ReplicateTensorToMesh, ConcatMeshToTensor, ListMeshToTensor
+
+
+import scipy
+from sklearn.metrics import top_k_accuracy_score
+import numpy as np
+
+from models.demos.t3000.llama2_70b.reference.llama.llama import Llama
+from models.demos.t3000.llama2_70b.tt.llama_generation import TtLlamaModelForGeneration
+
+
+from models.utility_functions import torch2tt_tensor, tt2torch_tensor, skip_for_grayskull, get_devices_for_t3000
+from models.demos.t3000.llama2_70b.tt.llama_common import (
+    setup_llama_env,
+    check_device_mesh,
+    extract_pcc_from_log,
+    MAX_SEQ_LEN,
+    BASE_URL,
+    UNIT_TEST_START_POS,
+    UNIT_TEST_GENERATION_LENGTH,
+    comp_pcc,
+    should_skip_model_load,
+    check_kv_cache,
+)
+
+from models.demos.t3000.llama2_70b.demo.demo import (
+    build_generator,
+    load_prompts_file,
+    intialize_inputs,
+    prepare_next_input,
+    get_sampling_func,
+    construct_arg,
+    DemoArgs,
+)
+
+
+def run_test_generation(args):
+    model_args = args.model
+    tt_args = args.tt
+    data_args = args.data
+    generator = Llama.build(
+        ckpt_dir=model_args.ckpt_dir,
+        tokenizer_path=model_args.tokenizer_path,
+        max_seq_len=model_args.max_seq_len,
+        max_batch_size=model_args.max_batch_size,
+        skip_model_load=model_args.skip_model_load,
+        n_layers=model_args.num_layers,
+    )
+
+    tt_model = TtLlamaModelForGeneration(
+        configuration=generator.model.params,
+        state_dict=generator.model.state_dict(),
+        model_args=model_args,
+        tt_args=tt_args,
+    )
+
+    pt_model = generator.model
+
+    tokenizer = generator.tokenizer
+    prompt_tokens, prompts = load_prompts_file(model_args, data_args, tokenizer)
+    sampling_func = get_sampling_func(data_args.top_k, data_args.top_p, data_args.temperature)
+
+    all_tests_pass = True
+    all_pccs, all_top1, all_top5 = [], [], []
+
+    # decode arguments
+    bsz = model_args.max_batch_size
+    max_gen_len = data_args.num_tokens
+
+    min_prompt_len = min(len(t) for t in prompt_tokens) if not tt_args.decode_only else 1
+    max_prompt_len = max(len(t) for t in prompt_tokens)
+    assert max_prompt_len <= model_args.max_kv_context_len
+    total_len = min(model_args.max_kv_context_len, max_gen_len + max_prompt_len)
+    assert total_len <= model_args.max_kv_context_len
```

Comment:
this logic has been repeated quite some times now. would it be better as a function?